README.md
---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->




## nhlmodelr

###Spatial Modelling of NHL Shot Location Data

Access pre-formatted NHL shot location data and implement spatially explicit models of NHL goal scoring probability. Advanced team, player, and goaltender statistics can be automatically calculated.

### Installation


```r
devtools::install_github("jedalong/nhlmodelr")
```
  
### Quick demo

Load  in some data (2011-12 season) and take a look at the structure.


```r
library(nhlmodelr)
data(shots20112012) 
head(shots20112012)
#>   PLAYID PLYR1.ID       PlyrName PlyrPos TeamID TeamStrength ShotType
#> 1    506      246   Patrik Elias      LW     11          701      805
#> 2    505     3343   David Krejci      C       1          701      803
#> 3    508      816 Bryce Salvador      D      11          701        0
#> 4    508     5490     Mark Fayne      D      11          701        0
#> 5    508     5430   Tyler Seguin      C       1          701        0
#> 6    507     5430   Tyler Seguin      C       1          701      802
#>   ShotDistFt GOALIE.ID     GoalieName gTeamID Goalie   x   y   SEASON
#> 1         14      1207     Tim Thomas       1    901 -76   1 20112012
#> 2         12       108 Martin Brodeur      11    901 -78   1 20112012
#> 3         NA        NA           <NA>       1      0 -70  -7 20112012
#> 4         NA        NA           <NA>       1      0 -42  24 20112012
#> 5         NA        NA           <NA>      11      0 -69 -18 20112012
#> 6         39        NA           <NA>      11    901 -51   0 20112012
#>   PreRegPost    GameID
#> 1          2 400046681
#> 2          2 400046681
#> 3          2 400046681
#> 4          2 400046681
#> 5          2 400046681
#> 6          2 400046681
summary(shots20112012)
#>      PLAYID         PLYR1.ID         PlyrName           PlyrPos         
#>  Min.   :505.0   Min.   :      0   Length:142668      Length:142668     
#>  1st Qu.:506.0   1st Qu.:   1576   Class :character   Class :character  
#>  Median :506.0   Median :   3127   Mode  :character   Mode  :character  
#>  Mean   :506.7   Mean   :  53534                                        
#>  3rd Qu.:508.0   3rd Qu.:   3792                                        
#>  Max.   :508.0   Max.   :2953322                                        
#>                                                                         
#>      TeamID       TeamStrength      ShotType       ShotDistFt    
#>  Min.   : 1.00   Min.   :  0.0   Min.   :  0.0   Min.   :  2.00  
#>  1st Qu.: 8.00   1st Qu.:701.0   1st Qu.:  0.0   1st Qu.: 18.00  
#>  Median :15.00   Median :701.0   Median :802.0   Median : 35.00  
#>  Mean   :15.31   Mean   :696.5   Mean   :598.7   Mean   : 35.89  
#>  3rd Qu.:23.00   3rd Qu.:701.0   3rd Qu.:803.0   3rd Qu.: 50.00  
#>  Max.   :30.00   Max.   :703.0   Max.   :807.0   Max.   :189.00  
#>                                                  NA's   :36294   
#>    GOALIE.ID        GoalieName           gTeamID          Goalie     
#>  Min.   :     72   Length:142668      Min.   : 1.00   Min.   :  0.0  
#>  1st Qu.:   1207   Class :character   1st Qu.: 8.00   1st Qu.:  0.0  
#>  Median :   2339   Mode  :character   Median :15.00   Median :901.0  
#>  Mean   :   2459                      Mean   :15.62   Mean   :672.1  
#>  3rd Qu.:   3502                      3rd Qu.:23.00   3rd Qu.:901.0  
#>  Max.   :2590376                      Max.   :30.00   Max.   :905.0  
#>  NA's   :66292                                                       
#>        x                y                SEASON           PreRegPost   
#>  Min.   :-99.00   Min.   :-42.0000   Min.   :20112012   Min.   :2.000  
#>  1st Qu.:-76.00   1st Qu.:-12.0000   1st Qu.:20112012   1st Qu.:2.000  
#>  Median :-64.00   Median :  1.0000   Median :20112012   Median :2.000  
#>  Mean   :-61.24   Mean   :  0.4313   Mean   :20112045   Mean   :2.068  
#>  3rd Qu.:-49.00   3rd Qu.: 13.0000   3rd Qu.:20112012   3rd Qu.:2.000  
#>  Max.   :  0.00   Max.   : 42.0000   Max.   :20122013   Max.   :3.000  
#>                                                                        
#>      GameID         
#>  Min.   :400046681  
#>  1st Qu.:400047008  
#>  Median :400047343  
#>  Mean   :400061787  
#>  3rd Qu.:400047671  
#>  Max.   :400265531  
#> 
```

Next, format the shots for the spatial model by spatially pooling all the shots to the 5525 unique locations in the offensive zone.


```r
modshots <- shots2model(shots20112012)
head(modshots)
#>     x   y xi ni
#> 1 -89 -42  0  0
#> 2 -88 -42  0  0
#> 3 -87 -42  0  0
#> 4 -86 -42  0  0
#> 5 -85 -42  0  0
#> 6 -84 -42  0  0
```

Set up parameters for the spatial model. First, we need to set the number of Gibbs samples for the MCMC chain and the burn-in. We also need to set-up initial values for each location (here we use random values between 0 and 1). The parameter p.max is the tuning parameter, which is the maximum probability of scoring a goal (in the centre of the goal). Finally, the output filename for storing output \*.RData file needs to be specified (the model produces a single \*.RData file with 8 objects, see  ?NHLmodelr).


```r
ngibbs <- 100
nburnin <- 50
p.max <- 0.29
inits <- runif(dim(modshots)[1],0,1)
outfile <- 'C:/Workspace/NHLModelOutput.RData'

#Model Run - Takes a couple of minutes
df <- NHLmodel(modshots,ngibbs,nburnin,p.max,inits,outfile)
head(df)
#>   x y oldx oldy newx newy      rad    theta index n g p_postmean
#> 1 1 1  -89  -42    0  -42 42.00000 1.570796     1 0 0 0.01120519
#> 2 2 1  -88  -42    1  -42 42.01190 1.546991     2 0 0 0.01235565
#> 3 3 1  -87  -42    2  -42 42.04759 1.523213     3 0 0 0.01244277
#> 4 4 1  -86  -42    3  -42 42.10701 1.499489     4 0 0 0.01488705
#> 5 5 1  -85  -42    4  -42 42.19005 1.475845     5 0 0 0.01560049
#> 6 6 1  -84  -42    5  -42 42.29657 1.452306     6 0 0 0.01731601
#>      p_postvar
#> 1 4.946174e-05
#> 2 5.825727e-05
#> 3 7.291214e-05
#> 4 1.577733e-04
#> 5 1.895014e-04
#> 6 1.961778e-04
```

Their are a number of plotting functions available. We can plot the raw shots data, only the goals, the posterior mean, and the posterior variance.


```r
NHLmodelplot(df,'shots')
```

![plot of chunk unnamed-chunk-5](README-unnamed-chunk-5-1.png)

```r
NHLmodelplot(df,'goals')
```

![plot of chunk unnamed-chunk-5](README-unnamed-chunk-5-2.png)

```r
#NHLmodelplot(df,'postmean')
#NHLmodelplot(df,'postvar')
```

We can also compute advanced team, player, and goaltender spatially adjusted scoring (and save percentage) statistics.


```r
#==== Not Run ====
# teamdb <- teams(df,shots20112012)
# playerdb <- players(df,shots20112012)
# goaliedb <- goalies(df,shots20112012)
#====
```

--- End ---


NEWS
nhlmodelr 0.1
=================
* 2018-02-20 - First attempt.


R/NHLmodelplot.R
# ---- roxygen documentation ----
#
#' @title Plot NHL model results
#'
#' @description
#'  Very basic plotting function for outputs from the spatial model of NHL goal scoring probabilities.
#'
#' @details
#'  Will be expanded later on.
#'
#' @param model a dataframe, specifically, ouput from the function \code{NHLmodel}.
#' @param plot character, what to plot; one of (\code{'shots','goals','postmean','postvar','none'}). If none is given just the rink outline is plotted.
#'
#' @return
#'  A half-rink plot of shots or goals (as points) or posterior mean or variance (as a surface).
#'
# @references
# @keywords
# @examples
#'
#' @export
#
# ---- End of roxygen documentation ----
NHLmodelplot <- function(model,plot='none'){
  ##############################################################################
  #plotting elements for the rink proper.
  
  faceoff.circle = function (x,y) {
    
    theta = seq(0,2*pi,length=300)
    #outer.
    polygon (x + 15*cos(theta),
             y + 15*sin(theta),
             lwd=2,
             border=2)
    polygon (x + 1*cos(theta),
             y + 1*sin(theta),
             col=2,
             border=2)
    segments (c(x-0.75,x-0.75, x+0.75,x+0.75, x-0.75,x-0.75, x+0.75,x+0.75),
              c(y-2,y-2, y-2,y-2, y+2,y+2,y+2,y+2),
              c(x-0.75,x-3.75, x+0.75,x+3.75, x-0.75,x-3.75, x+0.75,x+3.75),
              c(y-6,y-2, y-6,y-2, y+6,y+2,y+6,y+2),
              col=2, lwd=2)
    dd <- (5+7/12)/2
    segments (c(x-15, x-15, x+15, x+15),
              c(y-dd, y+dd, y-dd, y+dd),
              c(x-17, x-17, x+17, x+17),
              c(y-dd, y+dd, y-dd, y+dd),
              col=2, lwd=2)
  }
  
  goal.crease = function (flip=1) {
    xseq = seq(-4,4,length=100)
    polygon (c(-4, xseq, 4),
             flip*c(89, 83+xseq^2/4^2*1.5, 89),
             col="lightblue", border="red")
  }
  
  half.rink = function () {
    
    theta = seq(0,2*pi,length=300)
    #par(mar=c(0,0,0,0))
    plot(c(-42.6, 42.6), c(-101,0), ty="n", ylim=c(-101,0), xlim=c(-42.6, 42.6), ylab="", xlab="", axes=FALSE,asp=1)
    
    #Center Ice Circle
    polygon (15*cos(theta), 15*sin(theta), lwd=2, border=4)
    
    #Referee semi circle at penalty box
    theta2 = seq (pi/2, 3*pi/2, length=300)
    polygon (42.5 + 10*cos(theta2), 10*sin(theta2), lwd=2, border=2)
    
    #Center Ice
    rect(-42.5, -25, 42.5, -26, col=4, border=4)
    #Blue line
    rect(-42.5, -0.5, 42.5, 0.5, col=2, border=2)
    
    #Outside boards
    x <- 18
    lines (c(-42.5, 
             -42.5 + x - x*cos(seq(0,pi/2,length=20)),
             42.5 - x + x*cos(seq(pi/2,0,length=20)),
             42.5),
           c(15,
             -82 - x*sin(seq(0,pi/2,length=20)),
             -82 - x*sin(seq(pi/2,0,length=20)),
             15),
           col=1, lwd=2)
    
    
    goal.line.extreme = 42.5 - x + sqrt(x^2 - (x-11)^2)
    
    #the goal line & crese
    lines(goal.line.extreme*c(-1, 1), rep(-89,2), col=2,lwd=2)        #the goal line.
    lines(c(-3,-3,3,3), -(c(90,92,92,90)-1), col=1, lwd=3)    #the goal net.
    goal.crease(-1)
    ## traps.
    segments(c(-11, 11, -11, 11), c(89,89,-89,-89),
             c(-14,14,-14,14), c(100,100, -100,-100), col=2, lwd=2)
    
    faceoff.circle (22, -69)
    faceoff.circle (-22, -69)
    
    faceoff.dot = function (x,y) {
      polygon (x + 1*cos(theta),
               y + 1*sin(theta),
               col=2,
               border=2)
    }
    faceoff.dot (-22,-20)
    
    faceoff.dot (22,-20)
    
  }
  
  jitter <- function(x,y,nj,cs=0.5){
    jdf <- data.frame(x=runif(nj,min=x-cs,max=x+cs)) 
    jdf$y <- runif(nj, min=y-cs,max=y+cs)
    jdf
  }
  
  #Rotates x and y so that net is the bottom of plot, adjusted accordingly below.
  half.rink()
  
  #Plotting options
  # plotting all shots
  if (plot =='shots'){
    for (i in 1:dim(model)[1]){
      temp <- jitter(-1*model$oldy[i],model$oldx[i],model$n[i],0.5)   
      points(temp$x,temp$y,cex=0.05,pch=19)
    }
  }
  
  #Plotting goals
  if (plot == 'goals') {
    for (i in 1:dim(model)[1]){
      temp <- jitter(-1*model$oldy[i],model$oldx[i],model$g[i],0.5)   
      points(temp$x,temp$y,cex=0.05,pch=19)
    }
  }
  
  #plotting raster surface
  xx <- unique(model$oldx)
  yy <- unique(model$oldy)
  #col=rev(heat.colors(100,alpha=0.7))
  col=topo.colors(100,alpha=0.7)
  library(fields)
  
  #Plotting Posterior Mean
  if (plot == 'postmean'){
    mat <- matrix(data=model$p_postmean,nrow=length(xx),ncol=length(yy),byrow=F)
    mlim <- c(0,max(mat,na.rm=T))
    mat <- t(mat)  #Transpose to rotate plot
    image(yy,xx,mat,zlim=mlim,col=col,add=T)
    image.plot(legend.only=TRUE,zlim=mlim,col=col,legend.shrink=0.4,horizontal=F)
    contour(yy,xx,mat,levels=c(0.05,0.1,0.15,0.2),add=TRUE)
  }
  
  #Plotting Posterior variance
  if (plot == 'postvar'){
    mat <- matrix(data=model$p_postvar,nrow=length(xx),ncol=length(yy),byrow=F)
    mlim <- c(0,max(mat,na.rm=T))
    mat <- t(mat)
    image(yy,xx,mat,zlim=mlim,col=col,add=T)
    image.plot(legend.only=TRUE,zlim=mlim,col=col,legend.shrink=0.4,horizontal=F)
    contour(yy,xx,mat,levels=c(0.05,0.1,0.15,0.2),add=TRUE)
  }
  
}
